runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java

```diff
@@ -616,6 +649,186 @@ public void processElement(ProcessContext c) {
     p.run();
   }
 
+  /**
+   * Test running of {@link Pipeline} which has two {@link POutput POutputs} and finishing the first
+   * one triggers data being fed into the second one.
+   */
+  @Test(timeout = 10000)
+  public void testTwoPOutputsInPipelineWithCascade() throws InterruptedException {
+
+    StaticQueue<Integer> start = StaticQueue.of("start", VarIntCoder.of());
+    StaticQueue<Integer> messages = StaticQueue.of("messages", VarIntCoder.of());
+
+    Pipeline pipeline = getPipeline(false);
+    pipeline.begin().apply("outputStartSignal", outputStartTo(start));
+    PCollection<Integer> result =
+        pipeline
+            .apply("processMessages", messages.read())
+            .apply(
+                Window.<Integer>into(new GlobalWindows())
+                    .triggering(AfterWatermark.pastEndOfWindow())
+                    .discardingFiredPanes()
+                    .withAllowedLateness(Duration.ZERO))
+            .apply(Sum.integersGlobally());
+
+    // the result should be 6, after the data will have been written
+    PAssert.that(result).containsInAnyOrder(6);
+
+    PipelineResult run = pipeline.run();
+
+    // wait until a message has been written to the start queue
+    while (start.take() == null) {}
+
+    // and publish messages
+    messages.add(1).add(2).add(3).terminate();
+
+    run.waitUntilFinish();
+  }
+
+  @Test(timeout = 5000)
+  public void testTwoTimersSettingEachOther() {
+    Pipeline pipeline = getPipeline();
+    Instant now = new Instant(1500000000000L);
+    Instant end = now.plus(100);
+    PCollection<String> result = pipeline.apply(new TwoTimerTest(now, end));
```

Comment:
Reading this casually, `TwoTimerTest` is not really clear what is being tested. The code that creates the output and the code that asserts about the expectations are very far from each other. Ideally, you would move the assertion into `TwoTimerTest` somehow. Since `TwoTimerTest` owns the logic, it also owns the expected output.